crates/wasmprinter/src/operator.rs

```diff
@@ -1286,3 +1341,450 @@ impl<'a> VisitOperator<'a> for PrintOperator<'_, '_, '_, '_> {
+        }
+    }
+
+    // Set up the outermost block, representing the unnamed function label.
+    pub fn begin_function(&mut self, func_idx: u32) -> Result<()> {
+        match self.state.core.func_to_type.get(func_idx as usize) {
+            Some(Some(type_idx)) => self.control.push(Block {
+                ty: BlockType::FuncType(*type_idx),
+                kind: FrameKind::Block,
+                plain: String::new(),
+                folded: Vec::new(),
+                predicate: None,
+                consequent: None,
+                offset: self.operator_state.op_offset,
+            }),
+            _ => bail!("invalid func_idx"),
+        }
+
+        Ok(())
+    }
+
+    // Set up a catch-all block to represent the constant expression's operand stack.
+    pub fn begin_const_expr(&mut self) {
+        self.control.push(Block {
+            ty: BlockType::Empty,
+            kind: FrameKind::Block,
+            plain: String::new(),
+            folded: Vec::new(),
+            predicate: None,
+            consequent: None,
+            offset: 0,
+        });
+    }
+
+    // Handle a "plain" (non-block) instruction. The behavior resembles WABT's WatWriter::PushExpr().
+    // Each instruction will pop some number of operands off the stack to become the "children"
+    // of the foldedinst phrase. In the presence of multi-value instructions with more than 1 result,
+    // it may not be possible to hit the params target exactly. This doesn't necessarily mean the
+    // Wasm is invalid, but it can't be represented sensibly in folded form.
+    fn handle_plain(&mut self, plain: String, params: u32, mut results: u32) -> Result<()> {
+        match self.control.last_mut() {
+            Some(stack) => {
+                let mut first_param = stack.folded.len();
+                let mut param_count: u32 = 0;
+                if params > 0 {
+                    for (pos, inst) in stack.folded.iter().enumerate().rev() {
+                        param_count = param_count.saturating_add(inst.results);
+                        if param_count == params {
+                            first_param = pos;
+                            break;
+                        } else if param_count > params {
+                            // unfoldable instruction
+                            results = u32::MAX;
+                            break;
+                        }
+                    }
+                }
+
+                let mut inst = FoldedInstruction {
+                    plain,
+                    folded: stack.folded.drain(first_param..).collect(),
+                    results,
+                    offset: self.operator_state.op_offset,
+                };
+                if let Some(hint) = self.branch_hint.take() {
+                    inst.folded.push(hint);
+                }
+                stack.folded.push(inst);
+            }
+            _ => bail!("instruction without enclosing block"),
+        }
```

Comment:
As as a stylistic note one thing I like to do in situations like this to help reduce indentation is:

```rust
let stack = match self.control.last_mut() {
    Some(stack) => stack,
    None => bail!("..."),
};
// ...
```

that way it helps to avoid the nested indentation becoming too much.

Also, I'd generally recommend where possible to explicitly list out the "other cases" in a match rather than using `_`, for example here using `None` instead of `_`. For `Option<T>` it doesn't matter a whole lot because that'll never get another variant, but in general it helps with refactorings where if the match expression changes it's a good double-check to ensure each arm is updated as well

Reply:
Done.